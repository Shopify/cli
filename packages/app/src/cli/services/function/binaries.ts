import {joinPath, dirname} from '@shopify/cli-kit/node/path'
import {chmod, createFileWriteStream, fileExists, inTemporaryDirectory, mkdir, moveFile} from '@shopify/cli-kit/node/fs'
import {outputDebug} from '@shopify/cli-kit/node/output'
import {performActionWithRetryAfterRecovery} from '@shopify/cli-kit/common/retry'
import {fetch} from '@shopify/cli-kit/node/http'
import {PipelineSource} from 'stream'
import stream from 'node:stream/promises'
import fs from 'node:fs'
import * as gzip from 'node:zlib'
import {fileURLToPath} from 'node:url'

const FUNCTION_RUNNER_VERSION = 'v6.2.1'
const JAVY_VERSION = 'v3.1.0'

// The logic for determining the download URL and what to do with the response stream is _coincidentally_ the same for
// Javy and function-runner for now. Those methods may not continue to have the same logic in the future. If they
// diverge, make `Binary` an abstract class and create subclasses to handle the different logic polymorphically.
class DownloadableBinary {
  readonly name: string
  readonly version: string
  readonly path: string
  private readonly gitHubRepo: string

  constructor(name: string, version: string, gitHubRepo: string) {
    this.name = name
    this.version = version
    const filename = process.platform === 'win32' ? `${name}.exe` : name
    this.path = joinPath(dirname(fileURLToPath(import.meta.url)), '..', 'bin', filename)
    this.gitHubRepo = gitHubRepo
  }

  downloadUrl(processPlatform: string, processArch: string) {
    let platform
    let arch
    switch (processPlatform.toLowerCase()) {
      case 'darwin':
        platform = 'macos'
        break
      case 'linux':
        platform = 'linux'
        break
      case 'win32':
        platform = 'windows'
        break
      default:
        throw Error(`Unsupported platform ${processPlatform}`)
    }
    switch (processArch.toLowerCase()) {
      case 'arm':
      case 'arm64':
        arch = 'arm'
        break
      // A 32 bit arch likely needs that someone has 32bit Node installed on a
      // 64 bit system, and wasmtime doesn't support 32bit anyway.
      case 'ia32':
      case 'x64':
        arch = 'x86_64'
        break
      default:
        throw Error(`Unsupported architecture ${processArch}`)
    }

    const archPlatform = `${arch}-${platform}`
    // These are currently the same between both binaries _coincidentally_.
    const supportedTargets = ['arm-linux', 'arm-macos', 'x86_64-macos', 'x86_64-windows', 'x86_64-linux']
    if (!supportedTargets.includes(archPlatform)) {
      throw Error(`Unsupported platform/architecture combination ${processPlatform}/${processArch}`)
    }

    return `https://github.com/${this.gitHubRepo}/releases/download/${this.version}/${this.name}-${archPlatform}-${this.version}.gz`
  }

  async processResponse(responseStream: PipelineSource<unknown>, outputStream: fs.WriteStream): Promise<void> {
    const gunzip = gzip.createGunzip()
    await stream.pipeline(responseStream, gunzip, outputStream)
  }
}

let _javy: DownloadableBinary
let _functionRunner: DownloadableBinary

export function javyBinary() {
  if (!_javy) {
    _javy = new DownloadableBinary('javy', JAVY_VERSION, 'bytecodealliance/javy')
  }
  return _javy
}

export function functionRunnerBinary() {
  if (!_functionRunner) {
    _functionRunner = new DownloadableBinary('function-runner', FUNCTION_RUNNER_VERSION, 'Shopify/function-runner')
  }
  return _functionRunner
}

export async function installBinary(bin: DownloadableBinary) {
  const isInstalled = await fileExists(bin.path)
  if (isInstalled) {
    return
  }

  const url = bin.downloadUrl(process.platform, process.arch)
  outputDebug(`Downloading ${bin.name} ${bin.version} from ${url} to ${bin.path}`)
  const dir = dirname(bin.path)
  if (!(await fileExists(dir))) {
    await mkdir(dir)
  }
  await performActionWithRetryAfterRecovery(
    async () => {
      const resp = await fetch(url)
      if (resp.status !== 200) {
        throw new Error(`Downloading ${bin.name} failed with status code of ${resp.status}`)
      }

      const responseStream = resp.body
      if (responseStream === null) {
        throw new Error(`Downloading ${bin.name} failed with empty response body`)
      }

      // Download to a temp location and then move the file only after it's fully processed
      // so the `isInstalled` check above will continue to return false if the file hasn't
      // been fully processed.
      await inTemporaryDirectory(async (tmpDir) => {
        const tmpFile = joinPath(tmpDir, 'binary')
        const outputStream = createFileWriteStream(tmpFile)
        await bin.processResponse(responseStream, outputStream)
        await chmod(tmpFile, 0o775)
        await moveFile(tmpFile, bin.path)
      })
    },
    async () => {},
    2,
  )
}
