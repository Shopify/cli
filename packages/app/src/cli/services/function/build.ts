import {FunctionExtension} from '../../models/app/extensions.js'
import {exec} from '@shopify/cli-kit/node/system'
import {joinPath} from '@shopify/cli-kit/node/path'
import {build as esBuild} from 'esbuild'
import {findPathUp} from '@shopify/cli-kit/node/fs'
import {AbortSignal} from '@shopify/cli-kit/node/abort'
import {renderTasks} from '@shopify/cli-kit/node/ui'
import {Writable} from 'stream'

interface JSFunctionBuildOptions {
  stdout: Writable
  stderr: Writable
  signal?: AbortSignal
  // we want to use tasks when this is a primary command, i.e. 'shopify app function build',
  // but we don't want the fancy UI when this is running as part of 'shopify app build'.
  useTasks?: boolean
}

export async function buildJSFunction(fun: FunctionExtension, options: JSFunctionBuildOptions) {
  if (options.useTasks) {
    return buildJSFunctionWithTasks(fun, options)
  } else {
    return buildJSFunctionWithoutTasks(fun, options)
  }
}

async function buildJSFunctionWithoutTasks(fun: FunctionExtension, options: JSFunctionBuildOptions) {
  options.stdout.write(`Building GraphQL types...\n`)
  await buildGraphqlTypes(fun, options)
  options.stdout.write(`Bundling JS function...\n`)
  await bundleExtension(fun, options)
  options.stdout.write(`Running javy...\n`)
  await runJavy(fun, options)
  options.stdout.write(`Done!\n`)
}

export async function buildJSFunctionWithTasks(fun: FunctionExtension, options: JSFunctionBuildOptions) {
  return renderTasks([
    {
      title: 'Building GraphQL types',
      task: async () => {
        await buildGraphqlTypes(fun, options)
      },
    },
    {
      title: 'Bundling JS function',
      task: async () => {
        await bundleExtension(fun, options)
      },
    },
    {
      title: 'Running javy',
      task: async () => {
        await runJavy(fun, options)
      },
    },
  ])
}

export async function buildGraphqlTypes(
  fun: {directory: string; isJavaScript: boolean},
  options: JSFunctionBuildOptions,
) {
  if (!fun.isJavaScript) {
    throw new Error('GraphQL types can only be built for JavaScript functions')
  }

  return exec('npm', ['exec', '--', 'graphql-code-generator'], {
    cwd: fun.directory,
    stderr: options.stderr,
    signal: options.signal,
  })
}

export async function bundleExtension(fun: FunctionExtension, options: JSFunctionBuildOptions) {
  const entryPoint = await findPathUp('node_modules/@shopify/shopify_function/index.ts', {
    type: 'file',
    cwd: fun.directory,
  })
  if (!entryPoint) {
    throw new Error(
      "Could not find the Shopify Function runtime. Make sure you have '@shopify/shopify_function' installed",
    )
  }
  if (!fun.entrySourceFilePath) {
    throw new Error('Could not find your function entry point. It must be in src/index.js or src/index.ts')
  }

  const esbuildOptions = getESBuildOptions(fun.directory, entryPoint, fun.entrySourceFilePath)
  return esBuild(esbuildOptions)
}

function getESBuildOptions(directory: string, entryPoint: string, userFunction: string): Parameters<typeof esBuild>[0] {
  const esbuildOptions: Parameters<typeof esBuild>[0] = {
    outfile: joinPath(directory, 'dist/function.js'),
    entryPoints: [entryPoint],
    alias: {
      'user-function': userFunction,
    },
    logLevel: 'silent',
    bundle: true,
    legalComments: 'none',
    target: 'es2022',
    format: 'esm',
  }
  return esbuildOptions
}

export async function runJavy(fun: FunctionExtension, options: JSFunctionBuildOptions) {
  return exec('npm', ['exec', '--', 'javy', 'compile', '-o', fun.buildWasmPath, 'dist/function.js'], {
    cwd: fun.directory,
    stdout: options.stdout,
    stderr: options.stderr,
    signal: options.signal,
  })
}

export async function runFunctionRunner(fun: FunctionExtension) {
  return exec('npm', ['exec', '--', 'function-runner', '-f', fun.buildWasmPath], {
    cwd: fun.directory,
    stdin: 'inherit',
    stdout: 'inherit',
    stderr: 'inherit',
  })
}
