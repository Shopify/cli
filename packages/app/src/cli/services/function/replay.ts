/* eslint-disable @typescript-eslint/no-non-null-assertion */
import {renderReplay} from './ui.js'
import {runFunction} from './runner.js'
import {AppLinkedInterface} from '../../models/app/app.js'
import {ExtensionInstance} from '../../models/extensions/extension-instance.js'
import {FunctionConfigType} from '../../models/extensions/specifications/function.js'
import {selectFunctionRunPrompt} from '../../prompts/function/replay.js'

import {joinPath} from '@shopify/cli-kit/node/path'
import {readFile} from '@shopify/cli-kit/node/fs'
import {getLogsDir} from '@shopify/cli-kit/node/logs'
import {AbortError} from '@shopify/cli-kit/node/error'
import {AbortController} from '@shopify/cli-kit/node/abort'

import {readdirSync} from 'fs'

const LOG_SELECTOR_LIMIT = 100

interface ReplayOptions {
  app: AppLinkedInterface
  extension: ExtensionInstance<FunctionConfigType>
  stdout?: boolean
  path: string
  json: boolean
  watch: boolean
  log?: string
}

export interface FunctionRunData {
  shopId: number
  apiClientId: number
  payload: {
    input: unknown
    inputBytes: number
    output: unknown
    outputBytes: number
    functionId: string
    export: string
    logs: string
    fuelConsumed: number
  }
  logType: string
  cursor: string
  status: string
  source: string
  sourceNamespace: string
  logTimestamp: string
  identifier: string
}

export async function replay(options: ReplayOptions) {
  const {watch, extension, app} = options
  const abortController = new AbortController()

  try {
    const apiKey = options.app.configuration.client_id
    const functionRunsDir = joinPath(getLogsDir(), apiKey)

    const selectedRun = options.log
      ? await getRunFromIdentifier(functionRunsDir, extension.handle, options.log)
      : await getRunFromSelector(functionRunsDir, extension.handle)

    const {input, export: runExport} = selectedRun.payload

    if (watch) {
      await renderReplay({
        selectedRun,
        abortController,
        app,
        extension,
      })
    } else {
      await runFunction({
        functionExtension: extension,
        json: options.json,
        input: JSON.stringify(input),
        export: runExport,
      })
    }
  } catch (error) {
    abortController.abort()
    throw error
  }
}

async function getRunFromIdentifier(
  functionRunsDir: string,
  functionHandle: string,
  identifier: string,
): Promise<FunctionRunData> {
  const runPath = await findFunctionRun(functionRunsDir, functionHandle, identifier)
  if (runPath === undefined) {
    throw new AbortError(
      `No log found for '${identifier}'.\nSearched ${functionRunsDir} for function ${functionHandle}.`,
    )
  }
  const fileData = await readFile(runPath)
  return JSON.parse(fileData)
}

interface LogFileMetadata {
  namespace: string
  functionHandle: string
  identifier: string
}

function parseLogFilename(filename: string): LogFileMetadata | undefined {
  // Expected format: 20240522_150641_827Z_extensions_my-function_abcdef.json
  const splitFilename = filename.split(/[_.]/)

  if (splitFilename.length < 6) {
    return undefined
  } else {
    return {
      namespace: splitFilename[3]!,
      functionHandle: splitFilename[4]!,
      identifier: splitFilename[5]!,
    }
  }
}

async function findFunctionRun(
  functionRunsDir: string,
  functionHandle: string,
  identifier: string,
): Promise<string | undefined> {
  const fileName = readdirSync(functionRunsDir).find((filename) => {
    const fileMetadata = parseLogFilename(filename)
    return (
      fileMetadata?.namespace === 'extensions' &&
      fileMetadata?.functionHandle === functionHandle &&
      fileMetadata?.identifier === identifier
    )
  })

  return fileName ? joinPath(functionRunsDir, fileName) : undefined
}

async function getRunFromSelector(functionRunsDir: string, functionHandle: string): Promise<FunctionRunData> {
  const functionRuns = await getFunctionRunData(functionRunsDir, functionHandle)
  const selectedRun = await selectFunctionRunPrompt(functionRuns)

  if (selectedRun === undefined) {
    throw new AbortError(`No logs found in ${functionRunsDir}`)
  }
  return selectedRun
}

async function getFunctionRunData(functionRunsDir: string, functionHandle: string): Promise<FunctionRunData[]> {
  const allFunctionRunFileNames = readdirSync(functionRunsDir)
    .filter((filename) => {
      // Expected format: 20240522_150641_827Z_extensions_my-function_abcdef.json
      const fileMetadata = parseLogFilename(filename)
      return fileMetadata?.namespace === 'extensions' && fileMetadata?.functionHandle === functionHandle
    })
    .reverse()

  let functionRunData: FunctionRunData[] = []
  for (
    let i = 0;
    i < allFunctionRunFileNames.length && functionRunData.length < LOG_SELECTOR_LIMIT;
    i += LOG_SELECTOR_LIMIT
  ) {
    const currentFunctionRunFileNameChunk = allFunctionRunFileNames.slice(i, i + LOG_SELECTOR_LIMIT)
    const functionRunFilePaths = currentFunctionRunFileNameChunk.map((functionRunFile) =>
      joinPath(functionRunsDir, functionRunFile),
    )

    // eslint-disable-next-line no-await-in-loop
    const functionRunDataFromChunk = await Promise.all(
      functionRunFilePaths.map(async (functionRunFilePath) => {
        const fileData = await readFile(functionRunFilePath)
        const parsedData = JSON.parse(fileData)
        return {
          ...parsedData,
          identifier: getIdentifierFromFilename(functionRunFilePath),
        }
      }),
    )

    const filteredFunctionRunDataFromChunk = functionRunDataFromChunk.filter((run) => {
      return run.payload.input != null
    })

    functionRunData = functionRunData.concat(filteredFunctionRunDataFromChunk)
  }

  return functionRunData
}

function getIdentifierFromFilename(fileName: string): string {
  return fileName.split('_').pop()!.substring(0, 6)
}
