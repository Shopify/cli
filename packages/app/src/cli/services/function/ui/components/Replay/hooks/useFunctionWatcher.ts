import {FunctionRunData} from '../../../../replay.js'
import {AppInterface} from '../../../../../../models/app/app.js'
import {FunctionConfigType} from '../../../../../../models/extensions/specifications/function.js'
import {ExtensionInstance} from '../../../../../../models/extensions/extension-instance.js'
import {setupExtensionWatcher} from '../../../../../dev/extension/bundler.js'
import {FunctionRunFromRunner, ReplayLog} from '../types.js'
import {runFunction} from '../../../../runner.js'
import {AbortController} from '@shopify/cli-kit/node/abort'
import {useEffect, useState} from 'react'
import {useAbortSignal} from '@shopify/cli-kit/node/ui/hooks'
import {isUnitTest} from '@shopify/cli-kit/node/context/local'
import {treeKill} from '@shopify/cli-kit/node/tree-kill'
import {FatalError} from '@shopify/cli-kit/node/error'
import {Writable} from 'stream'

interface WatchFunctionForReplayOptions {
  selectedRun: FunctionRunData
  abortController: AbortController
  app: AppInterface
  extension: ExtensionInstance<FunctionConfigType>
}

export function useFunctionWatcher({selectedRun, abortController, app, extension}: WatchFunctionForReplayOptions) {
  const functionRunFromSelectedRun = {
    type: 'functionRun',
    input: selectedRun.payload.input,
    output: selectedRun.payload.output,
    logs: selectedRun.payload.logs,
    name: selectedRun.source,
    size: 0,
    memory_usage: 0,
    instructions: selectedRun.payload.fuelConsumed,
  } as FunctionRunFromRunner

  const [logs, setLogs] = useState<ReplayLog[]>([])
  const [recentFunctionRuns, setRecentFunctionRuns] = useState<[FunctionRunFromRunner, FunctionRunFromRunner]>([
    functionRunFromSelectedRun,
    functionRunFromSelectedRun,
  ])

  const [error, setError] = useState<string | undefined>(undefined)

  const {input, export: runExport} = selectedRun.payload

  const [statusMessage, setStatusMessage] = useState(`Watching for changes to ${selectedRun.source}...`)

  useEffect(() => {
    const watchAbortController = new AbortController()
    abortController.signal.addEventListener('abort', () => {
      watchAbortController.abort()
    })

    const runFunction = async () => {
      const functionRun = await runFunctionRunnerWithLogInput(extension, JSON.stringify(input), runExport)
      setRecentFunctionRuns((recentFunctionRuns) => {
        return [functionRun, recentFunctionRuns[0]]
      })
      setStatusMessage(`Watching for changes to ${selectedRun.source}...`)
      setLogs((logs) => [...logs, functionRun])
    }

    const initialReplay = async () => {
      setStatusMessage('Replaying log with local function...')
      await runFunction()
    }

    const startWatchingFunction = async () => {
      const customStdout = new Writable({
        write(chunk, _enconding, next) {
          setLogs((logs) => [...logs, {type: 'systemMessage', message: chunk.toString()}])
          next()
        },
      })

      await setupExtensionWatcher({
        extension,
        app,
        stdout: customStdout,
        stderr: customStdout,
        onChange: async () => {
          setError(undefined)
          setStatusMessage('Re-running with latest changes...')
          await runFunction()
        },
        onReloadAndBuildError: async (error) => {
          if (error instanceof FatalError) {
            setError(`Fatal error while reloading and building extension: ${error.formattedMessage || error.message}`)
          } else {
            setError(`Error while reloading and building extension: ${error.message}`)
          }
        },
        signal: watchAbortController.signal,
      })
    }

    // eslint-disable-next-line promise/catch-or-return, @typescript-eslint/no-floating-promises
    initialReplay().then(() => {
      // eslint-disable-next-line @typescript-eslint/no-floating-promises
      startWatchingFunction()
    })

    return () => {
      watchAbortController.abort()
    }
  }, [input, runExport, app, extension])

  useAbortSignal(abortController.signal, async () => {
    setTimeout(() => {
      if (isUnitTest()) return
      treeKill(process.pid, 'SIGINT', false, () => {
        process.exit(0)
      })
    }, 2000)
  })

  return {logs, statusMessage, recentFunctionRuns, error}
}

async function runFunctionRunnerWithLogInput(
  fun: ExtensionInstance<FunctionConfigType>,
  input: string,
  exportName: string,
): Promise<FunctionRunFromRunner> {
  let functionRunnerOutput = ''
  const customStdout = new Writable({
    write(chunk, _encoding, next) {
      functionRunnerOutput += chunk
      next()
    },
  })

  await runFunction({functionExtension: fun, input, export: exportName, stdout: customStdout, json: true})

  const result = JSON.parse(functionRunnerOutput)
  return {...result, type: 'functionRun'}
}
