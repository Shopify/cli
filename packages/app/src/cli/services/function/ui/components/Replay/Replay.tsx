import {FunctionRunFromRunner, ReplayLog} from './types.js'
import {useFunctionWatcher} from './hooks/useFunctionWatcher.js'
import {FunctionRunData} from '../../../replay.js'
import {ExtensionInstance} from '../../../../../models/extensions/extension-instance.js'
import {FunctionConfigType} from '../../../../../models/extensions/specifications/function.js'
import {AppLinkedInterface} from '../../../../../models/app/app.js'
import {prettyPrintJsonIfPossible} from '../../../../app-logs/utils.js'
import {AppEventWatcher} from '../../../../dev/app-events/app-event-watcher.js'
import figures from '@shopify/cli-kit/node/figures'
import {AbortController} from '@shopify/cli-kit/node/abort'
import React, {FunctionComponent} from 'react'
import {Box, Text, Static, useInput, useStdin} from '@shopify/cli-kit/node/ink'
import {useAbortSignal} from '@shopify/cli-kit/node/ui/hooks'
import {handleCtrlC} from '@shopify/cli-kit/node/ui'

export interface ReplayProps {
  selectedRun: FunctionRunData
  abortController: AbortController
  app: AppLinkedInterface
  extension: ExtensionInstance<FunctionConfigType>
}

const Replay: FunctionComponent<ReplayProps> = ({selectedRun, abortController, app, extension}) => {
  const {isAborted} = useAbortSignal(abortController.signal)
  const {isRawModeSupported: canUseShortcuts} = useStdin()
  const appWatcher = new AppEventWatcher(app)
  const {logs, statusMessage, recentFunctionRuns, error} = useFunctionWatcher({
    selectedRun,
    abortController,
    app,
    extension,
    appWatcher,
  })

  useInput(
    (input, key) => {
      handleCtrlC(input, key, () => abortController.abort())
      if (input === 'q') {
        abortController.abort()
      }
    },
    {isActive: Boolean(canUseShortcuts)},
  )

  return (
    <>
      {/* Scrolling upper section */}
      <Static items={logs}>
        {(log, index) => {
          return (
            <Box key={`replayOutputScrollerLog${index}`} flexDirection="column">
              <ReplayLog log={log} />
            </Box>
          )
        }}
      </Static>
      {/* Bottom Bar */}
      {/* eslint-disable-next-line no-negated-condition */}
      {!isAborted ? (
        <Box
          marginY={1}
          paddingTop={1}
          flexDirection="column"
          flexGrow={1}
          borderStyle="single"
          borderBottom={false}
          borderLeft={false}
          borderRight={false}
          borderTop
        >
          {canUseShortcuts ? (
            <Box flexDirection="column">
              <Box flexDirection="row">
                <Text>
                  {figures.pointerSmall} {statusMessage}
                </Text>
              </Box>
              <StatsDisplay recentFunctionRuns={recentFunctionRuns} />
              <Text>
                {figures.pointerSmall} Press <Text bold>q</Text> {figures.lineVertical} quit
              </Text>
            </Box>
          ) : null}
          {error ? <Text color="red">{error}</Text> : null}
        </Box>
      ) : null}
    </>
  )
}

function Header({title, backgroundColor}: {title: string; backgroundColor: string}) {
  const numSpaces = Math.max(Math.round((28 - title.length) / 2), 0)
  const spaces = ' '.repeat(numSpaces)
  return (
    <Text backgroundColor={backgroundColor} color="black">
      {'\n\n'}
      {spaces}
      {title}
      {spaces}
      {'\n'}
    </Text>
  )
}

function InputDisplay({input}: {input: string}) {
  return (
    <Box flexDirection="column">
      <Header title="Input" backgroundColor="yellow" />
      <Text>{prettyPrintJsonIfPossible(input)}</Text>
    </Box>
  )
}

function LogDisplay({logs}: {logs: string}) {
  return (
    <Box flexDirection="column">
      <Header title="Logs" backgroundColor="blueBright" />
      <Text>{logs}</Text>
    </Box>
  )
}

function OutputDisplay({output}: {output: string}) {
  return (
    <Box flexDirection="column">
      <Header title="Output" backgroundColor="greenBright" />
      <Text>{prettyPrintJsonIfPossible(output)}</Text>
    </Box>
  )
}

function BenchmarkDisplay({functionRun}: {functionRun: FunctionRunFromRunner}) {
  return (
    <Box flexDirection="column">
      <Header title="Benchmark Results" backgroundColor="green" />
      <Text>Name: {functionRun.name}</Text>
      <Text>Linear Memory Usage: {functionRun.memory_usage}KB</Text>
      <Text>Instructions: {functionRun.instructions / 1000}K</Text>
      <Text>Size: {functionRun.size}KB</Text>
    </Box>
  )
}

function StatsDisplay({recentFunctionRuns}: {recentFunctionRuns: [FunctionRunFromRunner, FunctionRunFromRunner]}) {
  const delta = recentFunctionRuns[0].instructions - recentFunctionRuns[1].instructions
  return (
    <Box flexDirection="column">
      <Text>
        {figures.pointerSmall} Instruction count change: {delta > 0 ? '+' : ''}
        {delta}
      </Text>
    </Box>
  )
}

function ReplayLog({log}: {log: ReplayLog}) {
  switch (log.type) {
    case 'functionRun':
      return (
        <Box flexDirection="column">
          <InputDisplay input={log.input} />
          <LogDisplay logs={log.logs} />
          <OutputDisplay output={log.output} />
          <BenchmarkDisplay functionRun={log} />
          <Text>&nbsp;</Text>
        </Box>
      )
    case 'systemMessage':
      return <Text>{log.message}</Text>
    default:
      return null
  }
}

export {Replay}
