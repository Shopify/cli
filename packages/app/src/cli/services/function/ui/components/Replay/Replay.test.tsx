import {Replay} from './Replay.js'
import {useFunctionWatcher} from './hooks/useFunctionWatcher.js'
import {FunctionRunFromRunner} from './types.js'
import {testFunctionExtension, testAppLinked} from '../../../../../models/app/app.test-data.js'
import {ExtensionInstance} from '../../../../../models/extensions/extension-instance.js'
import {FunctionConfigType} from '../../../../../models/extensions/specifications/function.js'
import {FunctionRunData} from '../../../replay.js'
import {AbortController} from '@shopify/cli-kit/node/abort'
import React from 'react'
import {beforeAll, describe, expect, test, vi} from 'vitest'
import {unstyled} from '@shopify/cli-kit/node/output'
import {render, sendInputAndWait, waitForInputsToBeReady} from '@shopify/cli-kit/node/testing/ui'

vi.mock('./hooks/useFunctionWatcher.js')

const SELECTED_RUN = {
  shopId: 69665030382,
  apiClientId: 124042444801,
  payload: {
    export: 'run',
    input: {
      someInput: 'someInput',
    },
    inputBytes: 136,
    output: {
      someOutput: 'someOutput',
    },
    outputBytes: 195,
    logs: 'First Log\nLog the second!,\n1,\nfourth line, length should be above!,\nFifth line!',
    functionId: '34236fa9-42f5-4bb6-adaf-956e12fff0b0',
    fuelConsumed: 532632,
  },
  logType: 'function_run',
  cursor: '2024-08-02T17:45:27.683139Z',
  status: 'success',
  source: 'product-discount',
  sourceNamespace: 'extensions',
  logTimestamp: '2024-08-02T17:45:27.382075Z',
  identifier: '123456',
} as FunctionRunData

const FUNCTION_RUN_FROM_SELECTED_RUN = {
  type: 'functionRun',
  input: SELECTED_RUN.payload.input,
  output: SELECTED_RUN.payload.output,
  logs: SELECTED_RUN.payload.logs,
  name: SELECTED_RUN.source,
  size: 0,
  memory_usage: 0,
  instructions: SELECTED_RUN.payload.fuelConsumed,
} as FunctionRunFromRunner

const FUNCTION_RUN_FROM_SELECTED_RUN2 = {
  type: 'functionRun',
  input: SELECTED_RUN.payload.input,
  output: SELECTED_RUN.payload.output,
  logs: SELECTED_RUN.payload.logs,
  name: SELECTED_RUN.source,
  size: 1,
  memory_usage: 1,
  instructions: SELECTED_RUN.payload.fuelConsumed + 1,
} as FunctionRunFromRunner

const WATCHER_RETURN_VALUE = {
  logs: [FUNCTION_RUN_FROM_SELECTED_RUN2, FUNCTION_RUN_FROM_SELECTED_RUN],
  isAborted: false,
  canUseShortcuts: true,
  statusMessage: `Watching for changes to ${SELECTED_RUN.source}...`,
  recentFunctionRuns: [FUNCTION_RUN_FROM_SELECTED_RUN2, FUNCTION_RUN_FROM_SELECTED_RUN],
  error: '',
}

let extension: ExtensionInstance<FunctionConfigType>

beforeAll(async () => {
  extension = await testFunctionExtension()
})

describe('Replay', () => {
  test('renders a stream of lines from function-runner output, and shortcuts', async () => {
    const mockedUseFunctionWatcher = vi.fn().mockReturnValue(WATCHER_RETURN_VALUE)
    vi.mocked(useFunctionWatcher).mockImplementation(mockedUseFunctionWatcher)

    const renderInstanceReplay = render(
      <Replay
        selectedRun={SELECTED_RUN}
        abortController={new AbortController()}
        app={testAppLinked()}
        extension={extension}
      />,
    )

    // Then
    expect(unstyled(renderInstanceReplay.lastFrame()!)).toMatchSnapshot()

    // unmount so that polling is cleared after every test
    renderInstanceReplay.unmount()
  })

  test('renders error in the bottom bar when present', async () => {
    const watcherReturnValueWithError = {
      ...WATCHER_RETURN_VALUE,
      error: 'some error',
    }
    const mockedUseFunctionWatcher = vi.fn().mockReturnValue(watcherReturnValueWithError)
    vi.mocked(useFunctionWatcher).mockImplementation(mockedUseFunctionWatcher)

    const renderInstanceReplay = render(
      <Replay
        selectedRun={SELECTED_RUN}
        abortController={new AbortController()}
        app={testAppLinked()}
        extension={extension}
      />,
    )

    // Then
    expect(unstyled(renderInstanceReplay.lastFrame()!)).toMatchSnapshot()

    // unmount so that polling is cleared after every test
    renderInstanceReplay.unmount()
  })

  test('quits when q is pressed', async () => {
    const abortController = new AbortController()
    const abortSpy = vi.spyOn(abortController, 'abort')

    const mockedUseFunctionWatcher = vi.fn().mockReturnValue(WATCHER_RETURN_VALUE)
    vi.mocked(useFunctionWatcher).mockImplementation(mockedUseFunctionWatcher)

    const renderInstanceReplay = render(
      <Replay
        selectedRun={SELECTED_RUN}
        abortController={abortController}
        app={testAppLinked()}
        extension={extension}
      />,
    )

    const promise = renderInstanceReplay.waitUntilExit()

    await waitForInputsToBeReady()
    await sendInputAndWait(renderInstanceReplay, 100, 'q')

    await promise

    // Then
    expect(abortSpy).toHaveBeenCalledOnce()

    // unmount so that polling is cleared after every test
    renderInstanceReplay.unmount()
  })

  test('quits when ctrl+c is pressed', async () => {
    // Given
    const abortController = new AbortController()
    const abortSpy = vi.spyOn(abortController, 'abort')

    const mockedUseFunctionWatcher = vi.fn().mockReturnValue(WATCHER_RETURN_VALUE)
    vi.mocked(useFunctionWatcher).mockImplementation(mockedUseFunctionWatcher)

    const renderInstanceReplay = render(
      <Replay
        selectedRun={SELECTED_RUN}
        abortController={abortController}
        app={testAppLinked()}
        extension={extension}
      />,
    )

    const promise = renderInstanceReplay.waitUntilExit()

    await waitForInputsToBeReady()
    await sendInputAndWait(renderInstanceReplay, 100, '\u0003')

    await promise
    // Then
    expect(abortSpy).toHaveBeenCalledOnce()

    // unmount so that polling is cleared after every test
    renderInstanceReplay.unmount()
  })
})
