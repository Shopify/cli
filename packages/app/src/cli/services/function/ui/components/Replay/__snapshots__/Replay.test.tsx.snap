// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Replay > renders a stream of lines from function-runner output, and shortcuts 1`] = `
"

            Input            

{
  \\"someInput\\": \\"someInput\\"
}


            Logs            

First Log
Log the second!,
1,
fourth line, length should be above!,
Fifth line!


           Output           

{
  \\"someOutput\\": \\"someOutput\\"
}


      Benchmark Results      

Name: product-discount
Linear Memory Usage: 1KB
Instructions: 532.633K
Size: 1KB



            Input            

{
  \\"someInput\\": \\"someInput\\"
}


            Logs            

First Log
Log the second!,
1,
fourth line, length should be above!,
Fifth line!


           Output           

{
  \\"someOutput\\": \\"someOutput\\"
}


      Benchmark Results      

Name: product-discount
Linear Memory Usage: 0KB
Instructions: 532.632K
Size: 0KB


────────────────────────────────────────────────────────────────────────────────────────────────────

› Watching for changes to product-discount...
› Instruction count change: +1
› Press q │ quit
"
`;

exports[`Replay > renders error in the bottom bar when present 1`] = `
"

            Input            

{
  \\"someInput\\": \\"someInput\\"
}


            Logs            

First Log
Log the second!,
1,
fourth line, length should be above!,
Fifth line!


           Output           

{
  \\"someOutput\\": \\"someOutput\\"
}


      Benchmark Results      

Name: product-discount
Linear Memory Usage: 1KB
Instructions: 532.633K
Size: 1KB



            Input            

{
  \\"someInput\\": \\"someInput\\"
}


            Logs            

First Log
Log the second!,
1,
fourth line, length should be above!,
Fifth line!


           Output           

{
  \\"someOutput\\": \\"someOutput\\"
}


      Benchmark Results      

Name: product-discount
Linear Memory Usage: 0KB
Instructions: 532.632K
Size: 0KB


────────────────────────────────────────────────────────────────────────────────────────────────────

› Watching for changes to product-discount...
› Instruction count change: +1
› Press q │ quit
some error
"
`;
